// catalog-converter/dist/lib/converters/Wps.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var dist_1 = __importDefault(require("@sindresorhus/is/dist"));
var helpers_1 = require("./helpers");
var Message_1 = require("../Message");
var lodash_1 = require("lodash");
function wpsCatalogItem(item, options) {
    if (!options.partial && !dist_1.default.string(item.url)) {
        return helpers_1.nullResult(Message_1.missingRequiredProp(Message_1.ModelType.WpsItem, "url", "string", item.name));
    }
    var propsToCopy = [
        "url",
        "identifier",
        "description",
        "executeWithHttpGet",
    ];
    var unknownProps = helpers_1.getUnknownProps(item, __spreadArrays(helpers_1.catalogMemberProps, helpers_1.catalogMemberPropsIgnore, propsToCopy, [
        "featureInfoTemplate",
    ]));
    var member = {
        type: "wps",
        name: item.name,
    };
    var messages = helpers_1.propsToWarnings(Message_1.ModelType.WpsItem, unknownProps, item.name);
    helpers_1.copyProps(item, member, __spreadArrays(helpers_1.catalogMemberProps, propsToCopy));
    if (options.copyUnknownProperties) {
        helpers_1.copyProps(item, member, unknownProps);
    }
    return {
        member: member,
        messages: messages,
    };
}
exports.wpsCatalogItem = wpsCatalogItem;
function wpsResultItem(item, options) {
    var propsToCopy = ["wpsResponseUrl", "wpsResponse", "parameters"];
    // do something for parameterValues
    var unknownProps = helpers_1.getUnknownProps(item, __spreadArrays(helpers_1.catalogMemberProps, helpers_1.catalogMemberPropsIgnore, propsToCopy, [
        "featureInfoTemplate",
    ]));
    var member = {
        type: "wps-result",
        name: item.name,
    };
    var messages = helpers_1.propsToWarnings(Message_1.ModelType.WpsResultItem, unknownProps, item.name);
    helpers_1.copyProps(item, member, __spreadArrays(helpers_1.catalogMemberProps, propsToCopy));
    if (options.copyUnknownProperties) {
        helpers_1.copyProps(item, member, unknownProps);
    }
    return {
        member: member,
        messages: messages,
    };
}
exports.wpsResultItem = wpsResultItem;
function wpsCatalogGroup(item, options) {
    var error;
    if (!dist_1.default.string(item.url)) {
        error = Message_1.missingRequiredProp(Message_1.ModelType.WpsGroup, "url", "string", item.name);
    }
    if (!options.partial && error) {
        return {
            member: null,
            messages: [error],
        };
    }
    var propsToCopy = ["isOpen"];
    var unknownProps = helpers_1.getUnknownProps(item, __spreadArrays(helpers_1.catalogMemberProps, helpers_1.catalogMemberPropsIgnore, propsToCopy, [
        "itemProperties",
    ]));
    var member = {
        type: "wps-getCapabilities",
        name: item.name,
    };
    var messages = helpers_1.propsToWarnings(Message_1.ModelType.WpsGroup, unknownProps, item.name);
    if (options.copyUnknownProperties) {
        helpers_1.copyProps(item, member, unknownProps);
    }
    helpers_1.copyProps(item, member, __spreadArrays(helpers_1.catalogMemberProps, propsToCopy));
    if (lodash_1.isPlainObject(item.itemProperties)) {
        var itemPropertiesResult = helpers_1.itemProperties(item, wpsCatalogItem, options);
        if (itemPropertiesResult.result)
            member.itemProperties = itemPropertiesResult.result;
        messages.push.apply(messages, itemPropertiesResult.messages);
    }
    return { member: member, messages: messages };
}
exports.wpsCatalogGroup = wpsCatalogGroup;
