// catalog-converter/dist/lib/Message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var opaque_type_1 = require("@iadvize-oss/opaque-type");
var foldable_helpers_1 = require("@iadvize-oss/foldable-helpers");
var Severity;
(function (Severity) {
    Severity[Severity["Error"] = 0] = "Error";
    Severity[Severity["Warning"] = 1] = "Warning";
})(Severity = exports.Severity || (exports.Severity = {}));
var ModelType;
(function (ModelType) {
    ModelType["Share"] = "Share";
    ModelType["Member"] = "CatalogMember";
    ModelType["Group"] = "CatalogGroup";
    ModelType["WmsItem"] = "WebMapServiceCatalogItem";
    ModelType["WmsGroup"] = "WebMapServiceCatalogGroup";
    ModelType["WfsGroup"] = "WebFeatureServiceCatalogGroup";
    ModelType["CsvItem"] = "CsvCatalogItem";
    ModelType["SosItem"] = "SensorObservationServiceCatalogItem";
    ModelType["CswCatalogGroup"] = "CswCatalogGroup";
    ModelType["EsriCatalogGroup"] = "EsriCatalogGroup";
    ModelType["EsriMapServerGroup"] = "EsriMapServerGroup";
    ModelType["EsriMapServerItem"] = "EsriMapServerCatalogItem";
    ModelType["EsriFeatureServerItem"] = "EsriFeatureServerCatalogItem";
    ModelType["WpsItem"] = "WpsItem";
    ModelType["WpsResultItem"] = "WpsResultItem";
    ModelType["WpsGroup"] = "WpsGroup";
    ModelType["CkanGroup"] = "CkanCatalogGroup";
    ModelType["CkanCatalogItem"] = "CkanCatalogItem";
    ModelType["GeoJsonItem"] = "GeoJsonCatalogItem";
    ModelType["CartoMapCatalogItem"] = "CartoMapCatalogItem";
    ModelType["MapboxVectorTileCatalogItem"] = "MapboxVectorTileCatalogItem";
    ModelType["KmlCatalogItem"] = "KmlCatalogItem";
    ModelType["SocrataCatalogGroup"] = "SocrataCatalogGroup";
})(ModelType = exports.ModelType || (exports.ModelType = {}));
exports.unknownPropOpaque = opaque_type_1.createOpaqueAPI("UnknownProp");
function isUnknownProp(m) {
    return exports.unknownPropOpaque.isOpaque(m.details);
}
exports.isUnknownProp = isUnknownProp;
var missingRequiredPropOpaque = opaque_type_1.createOpaqueAPI("MissingRequiredProp");
function isMissingRequiredProp(m) {
    return missingRequiredPropOpaque.isOpaque(m.details);
}
exports.isMissingRequiredProp = isMissingRequiredProp;
var unknownTypeOpaque = opaque_type_1.createOpaqueAPI("UnknownType");
function isUnknownType(m) {
    return unknownTypeOpaque.isOpaque(m.details);
}
exports.isUnknownType = isUnknownType;
var inputNotPlainObjectOpaque = opaque_type_1.createOpaqueAPI("InputNotPlainObject");
function isInputNotPlainObject(m) {
    return inputNotPlainObjectOpaque.isOpaque(m.details);
}
exports.isInputNotPlainObject = isInputNotPlainObject;
function unknownProp(modelType, property, label, severity) {
    if (severity === void 0) { severity = Severity.Warning; }
    return {
        message: "Unknown property \"" + property + "\" was encountered",
        path: [label],
        severity: severity,
        details: exports.unknownPropOpaque.toOpaque({ modelType: modelType, property: property }),
    };
}
exports.unknownProp = unknownProp;
function missingRequiredProp(modelType, property, requiredType, label, severity) {
    if (severity === void 0) { severity = Severity.Error; }
    return {
        message: modelType + " missing \"" + (typeof property === "string" ? property : property.join('" or "')) + "\"",
        path: [label],
        severity: severity,
        details: missingRequiredPropOpaque.toOpaque({
            modelType: modelType,
            property: property,
            requiredType: requiredType,
        }),
    };
}
exports.missingRequiredProp = missingRequiredProp;
function unknownType(type, label, severity) {
    if (severity === void 0) { severity = Severity.Error; }
    return {
        message: "Unknown or unsupported type \"" + type + "\"",
        path: [label],
        severity: severity,
        details: unknownTypeOpaque.toOpaque({ type: type }),
    };
}
exports.unknownType = unknownType;
function inputNotPlainObject() {
    return {
        message: "Input to conversion functions must be a plain object",
        path: [],
        severity: Severity.Error,
        details: inputNotPlainObjectOpaque.toOpaque({}),
    };
}
exports.inputNotPlainObject = inputNotPlainObject;
function getUnknownPropDetails(m) {
    return exports.unknownPropOpaque.fromOpaque(m.details);
}
exports.getUnknownPropDetails = getUnknownPropDetails;
function getMissingRequiredPropDetails(m) {
    return missingRequiredPropOpaque.fromOpaque(m.details);
}
exports.getMissingRequiredPropDetails = getMissingRequiredPropDetails;
function getUnknownTypeDetails(m) {
    return unknownTypeOpaque.fromOpaque(m.details);
}
exports.getUnknownTypeDetails = getUnknownTypeDetails;
function getInputNotPlainObjectDetails(m) {
    return inputNotPlainObjectOpaque.fromOpaque(m.details);
}
exports.getInputNotPlainObjectDetails = getInputNotPlainObjectDetails;
exports.foldMessage = foldable_helpers_1.createFoldObject({
    isUnknownProp: isUnknownProp,
    isUnknownType: isUnknownType,
    isMissingRequiredProp: isMissingRequiredProp,
    isInputNotPlainObject: isInputNotPlainObject,
});
